security.md

```diff
@@ -0,0 +1,94 @@
+clients if desired.
+
+## Denial of Service (DoS) attacks on the Rendezvous server
+
+Wormhole codes can be so short because they implicitly contain a common
+rendezvous server URL (any two applications that use Magic Wormhole
+should be configured to use the same server). As a result, successful
+operation depends upon both clients being able to contact that server,
+making it a single point of failure (SPOF).
+
+While it is of course possible to incapacitate the rendezvous server using
+traditional means of DoS, so much work is not necessary: An attacker may
+simply connect to every nameplate with a random code, causing the key exchange
+to fail. There is a "list" command in the protocol which makes it easy to
+enumerate all nameplates to disrupt them, but even without it the name space is
+sufficiently small (by design, as we want short codes) to brute force.
+
+There are several ideas to defend against this, however the only one already
+specified in the protocol is the "permission" feature: The rendezvous server
+may use it to restrict new connections with some kind of challenge (or for
+private servers, authentication). The motivation behind this is to enable
+a proof-of-work challenge (using HashCash) when under attack to slow it down.
+See the server protocol specification for details.
+
+Currently the best defense against this is to use high-entropy nameplates which
+cannot easily be enumerated at the cost of longer codes. This has not be
+implemented yet though, and would still leave the attack vector of "ordinary"
+DoS.
+
+## Metadata leaks
+
+As usual in cryptography, communication metadata cannot easily be protected. Any
+involved server will know the timestamp and size of each message it relays. This
+usually is not much of an issue, except for file transfer where the file size may
+be characteristic for its content. (Imagine people using Magic Wormhole to send
+some specific movie.)
+
+Thanks to the separation between rendezvous server and relay server, forcing the
+client to establish a direct connection will bypass the relay server and thus not
+leak such metadata for file transfers.
+
+## IP address leakage and anonymity
+
+As usual on the internet, any involved server will know connection information
+including the IP address of any two peers.
+
+Additionally, in order to establish a direct connection, public and local IP
+addresses are sent to the peer. This may leak information about the local
+network topology, including potentially used VPNs. If you do not trust your
+peer with such information, enforcing a relay server connection is advised.
+However, keep in mind that it is easy to host a relay server, so it might still
+leak your public IP address.
+
+If you do not trust the servers with knowing your IP address either, use
+[Tor](https://torproject.org/) to use Magic Wormhole anonymously. The Transit
+protocol has Tor support built-in, which then makes use of the Tor network itself
+as the relay.
```

Comment:
I'm not clear what you mean here? The transit is still involved, because we lack support for Onion services (this part is mentioned in a file you've deleted -- perhaps just including that stuff here would be better).

Reply:
I was referring to the Tor part which should be documented in the Transit protocol specification. I'll make it more explicit

Reply:
It's still using the transit server though, so the part about "use the Tor network itself as the relay" isn't really correct (or I don't understand what you mean here).

Reply:
Yes you can always connect to a relay from behind Tor, but that's not what I'm talking about. The transit protocol specifies "Tor hints" which point to onion domains, of which the idea is to spin up a hidden service and send that over as a hint. In that case, there is a "direct" connection with respect to the Tor network. But from an IP perspective, the Tor network merely acts as a relay.

Reply:
Oh, it actually spins up Onion services?? Cool .. I guess it's been a while since I looked in there very deeply.

I suppose you could consider that Tor is basically acting "as" the relay in that case -- although I guess you could always say Tor is "a relay" in that sense. You could also view those as "merely" direct connections (albeit via Tor) since the semantics of Onion domains give you assurance you're connecting to the right computer.

Okay, I guess it's fine as-is.

It might still be worth distinguishing between "use Tor, but still via the Transit" vs. "use only Tor"...? (The former thing is possible, right?)

Reply:
> Oh, it actually spins up Onion services?

Actually I don't know ^^ I know for sure that the protocol supports this and that this definitely was the original plan. But to which extent it has been implemented in the Python client, no idea. (IMO the latter question does not matter that much because this document should mostly speak from a protocol perspective.)

> It might still be worth distinguishing between "use Tor, but still via the Transit" vs. "use only Tor"...? (The former thing is possible, right?)

Yes, I'll try to work that in